// graphql-transformer-common/lib/ResourceConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ResourceConstants = /** @class */ (function () {
    function ResourceConstants() {
    }
    ResourceConstants.NONE = "NONE";
    ResourceConstants.RESOURCES = {
        // AppSync
        GraphQLAPILogicalID: 'GraphQLAPI',
        GraphQLSchemaLogicalID: 'GraphQLSchema',
        APIKeyLogicalID: 'GraphQLAPIKey',
        // ElasticSearch
        ElasticSearchAccessIAMRoleLogicalID: 'ElasticSearchAccessIAMRole',
        ElasticSearchDomainLogicalID: 'ElasticSearchDomain',
        ElasticSearchStreamingLambdaIAMRoleLogicalID: 'ElasticSearchStreamingLambdaIAMRole',
        ElasticSearchStreamingLambdaFunctionLogicalID: 'ElasticSearchStreamingLambdaFunction',
        ElasticSearchDataSourceLogicalID: 'ElasticSearchDataSource',
        // Auth
        AuthCognitoUserPoolLogicalID: 'AuthCognitoUserPool',
        AuthCognitoUserPoolNativeClientLogicalID: 'AuthCognitoUserPoolNativeClient',
        AuthCognitoUserPoolJSClientLogicalID: 'AuthCognitoUserPoolJSClient',
    };
    ResourceConstants.PARAMETERS = {
        // AppSync
        AppSyncApiName: 'AppSyncApiName',
        APIKeyExpirationEpoch: 'APIKeyExpirationEpoch',
        // DynamoDB
        DynamoDBModelTableReadIOPS: 'DynamoDBModelTableReadIOPS',
        DynamoDBModelTableWriteIOPS: 'DynamoDBModelTableWriteIOPS',
        // ElasticSearch
        ElasticSearchDomainName: 'ElasticSearchDomainName',
        ElasticSearchAccessIAMRoleName: 'ElasticSearchAccessIAMRoleName',
        ElasticSearchDebugStreamingLambda: 'ElasticSearchDebugStreamingLambda',
        ElasticSearchStreamingIAMRoleName: 'ElasticSearchStreamingIAMRoleName',
        ElasticSearchStreamingFunctionName: 'ElasticSearchStreamingFunctionName',
        ElasticSearchInstanceCount: 'ElasticSearchInstanceCount',
        ElasticSearchInstanceType: 'ElasticSearchInstanceType',
        ElasticSearchEBSVolumeGB: 'ElasticSearchEBSVolumeGB',
        ElasticSearchStreamingLambdaCodeS3Bucket: 'ElasticSearchStreamingLambdaCodeS3Bucket',
        ElasticSearchStreamingLambdaCodeS3Key: 'ElasticSearchStreamingLambdaCodeS3Key',
        ElasticSearchStreamingLambdaHandlerName: 'ElasticSearchStreamingLambdaHandlerName',
        ElasticSearchStreamingLambdaRuntime: 'ElasticSearchStreamingLambdaRuntime',
        // Auth
        AuthCognitoUserPoolId: 'AuthCognitoUserPoolId',
        AuthCognitoUserPoolName: 'AuthCognitoUserPoolName',
        AuthCognitoUserPoolMobileClientName: 'AuthCognitoUserPoolMobileClientName',
        AuthCognitoUserPoolJSClientName: 'AuthCognitoUserPoolJSClientName',
        AuthCognitoUserPoolRefreshTokenValidity: 'AuthCognitoUserPoolRefreshTokenValidity'
    };
    ResourceConstants.MAPPINGS = {};
    ResourceConstants.CONDITIONS = {
        // General
        APIKeyExpirationEpochIsZero: 'APIKeyExpirationEpochIsZero',
        APIKeyExpirationEpochIsNotNegOne: 'APIKeyExpirationEpochIsNotNegOne',
        APIKeyExpirationEpochIsPositive: 'APIKeyExpirationEpochIsPositive',
        // Auth
        AuthShouldCreateUserPool: 'AuthShouldCreateUserPool'
    };
    ResourceConstants.OUTPUTS = {
        // AppSync
        GraphQLAPIEndpointOutput: 'GraphQLAPIEndpointOutput',
        GraphQLAPIApiKeyOutput: 'GraphQLAPIKeyOutput',
        GraphQLAPIIdOutput: 'GraphQLAPIIdOutput',
        // Elasticsearch
        ElasticSearchStreamingLambdaIAMRoleArn: 'ElasticSearchStreamingLambdaIAMRoleArn',
        ElasticSearchAccessIAMRoleArn: 'ElasticSearchAccessIAMRoleArn',
        // Auth
        AuthCognitoUserPoolIdOutput: 'AuthCognitoUserPoolIdOutput',
        AuthCognitoUserPoolNativeClientOutput: 'AuthCognitoUserPoolNativeClientId',
        AuthCognitoUserPoolJSClientOutput: 'AuthCognitoUserPoolJSClientId'
    };
    ResourceConstants.METADATA = {};
    ResourceConstants.SNIPPETS = {
        AuthCondition: "authCondition",
        VersionedCondition: "versionedCondition",
        IsDynamicGroupAuthorizedVariable: "isDynamicGroupAuthorized",
        IsLocalDynamicGroupAuthorizedVariable: "isLocalDynamicGroupAuthorized",
        IsStaticGroupAuthorizedVariable: "isStaticGroupAuthorized",
        IsOwnerAuthorizedVariable: "isOwnerAuthorized",
        IsLocalOwnerAuthorizedVariable: "isLocalOwnerAuthorized"
    };
    return ResourceConstants;
}());
exports.ResourceConstants = ResourceConstants;
