// graphql-appsync-transformer/lib/resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var appSync_1 = require("cloudform/types/appSync");
var cloudform_1 = require("cloudform");
var graphql_transformer_common_1 = require("graphql-transformer-common");
var ResourceFactory = /** @class */ (function () {
    function ResourceFactory() {
    }
    ResourceFactory.prototype.makeResolverS3RootParams = function () {
        return {
            Parameters: {
                ResolverBucket: new cloudform_1.StringParameter({
                    Description: "The name of the bucket containing the resolver templates",
                }),
                ResolverRootKey: new cloudform_1.StringParameter({
                    Description: "The s3 key of the folder containing the resolver templates in format {Type}.{Field}.[response|request].{Timestamp}",
                }),
                DeploymentTimestamp: new cloudform_1.StringParameter({
                    Description: "The timestamp used to identify thie most recent version of the resolver templates in s3.",
                })
            }
        };
    };
    ResourceFactory.prototype.makeResolverParam = function (name) {
        var _a;
        return {
            Parameters: (_a = {},
                _a[this.removeDotsAndCamelcase(name)] = new cloudform_1.StringParameter({
                    Description: "The S3 location for the Resolver: " + name,
                }),
                _a)
        };
    };
    ResourceFactory.prototype.updateResolverResource = function (resource) {
        return new appSync_1.default.Resolver({
            ApiId: resource.Properties.ApiId,
            DataSourceName: resource.Properties.DataSourceName,
            FieldName: resource.Properties.FieldName,
            TypeName: resource.Properties.TypeName,
            RequestMappingTemplateS3Location: cloudform_1.Fn.Join('', [
                's3://',
                cloudform_1.Fn.Join('/', [
                    cloudform_1.Fn.Ref('ResolverBucket'),
                    cloudform_1.Fn.Ref('ResolverRootKey'),
                    cloudform_1.Fn.Join('.', [
                        resource.Properties.TypeName,
                        resource.Properties.FieldName,
                        'request',
                        cloudform_1.Fn.Ref('DeploymentTimestamp')
                    ])
                ]),
            ]),
            ResponseMappingTemplateS3Location: cloudform_1.Fn.Join('', [
                's3://',
                cloudform_1.Fn.Join('/', [
                    cloudform_1.Fn.Ref('ResolverBucket'),
                    cloudform_1.Fn.Ref('ResolverRootKey'),
                    cloudform_1.Fn.Join('.', [
                        resource.Properties.TypeName,
                        resource.Properties.FieldName,
                        'response',
                        cloudform_1.Fn.Ref('DeploymentTimestamp')
                    ])
                ]),
            ])
        }).dependsOn(graphql_transformer_common_1.ResourceConstants.RESOURCES.GraphQLSchemaLogicalID);
    };
    ResourceFactory.prototype.makeSchemaParam = function () {
        var _a;
        return {
            Parameters: (_a = {},
                _a[this.removeDotsAndCamelcase('schema.graphql')] = new cloudform_1.StringParameter({
                    Description: "The S3 location for the Schema: schema.graphql",
                }),
                _a)
        };
    };
    ResourceFactory.prototype.makeAppSyncSchema = function (schema) {
        if (schema) {
            return new appSync_1.default.GraphQLSchema({
                ApiId: cloudform_1.Fn.GetAtt(graphql_transformer_common_1.ResourceConstants.RESOURCES.GraphQLAPILogicalID, 'ApiId'),
                Definition: schema
            });
        }
        return new appSync_1.default.GraphQLSchema({
            ApiId: cloudform_1.Fn.GetAtt(graphql_transformer_common_1.ResourceConstants.RESOURCES.GraphQLAPILogicalID, 'ApiId'),
            DefinitionS3Location: cloudform_1.Fn.Ref(this.removeDotsAndCamelcase('schema.graphql'))
        });
    };
    ResourceFactory.prototype.removeDotsAndCamelcase = function (name) {
        var nameCopy = name;
        for (var i = 0; i < name.length; i++) {
            if (name[i] === '.') {
                nameCopy = nameCopy.substr(0, i + 1) + nameCopy.charAt(i + 1).toUpperCase() + nameCopy.slice(i + 2);
            }
        }
        return nameCopy.replace(/\./g, '');
    };
    return ResourceFactory;
}());
exports.ResourceFactory = ResourceFactory;
